File: 1/2021.10.08/187.Repeated DNA Sequences/python/solution.py
class Solution:
    def findRepeatedDnaSequences(self, s: str) -> List[str]:
        L = 10
        bin = {'A': 0, 'C': 1, 'G': 2, 'T': 3}
        n = len(s)
        if n <= L:
            return []
        res = []
        x = 0
        for ch in s[:L-1]:
            x = (x << 2) | bin[ch]
        cnt = defaultdict(int)
        # i + 10 - 1 == n - 1
        # i = n - 10
        for i in range(n - L + 1):
            x = ((x << 2) | bin[s[i + L - 1]]) & ((1 << L * 2) - 1)
            cnt[x] += 1
            if cnt[x] == 2:
                res.append(s[i:i + L])
        return res

File: 1/2021.10.08/187.Repeated DNA Sequences/java/solution.java
class Solution {
    public List<String> findRepeatedDnaSequences(String s) {
        if (s.length() < 11) {
            return new ArrayList<>();
        }
        Map<Character, Integer> map = new HashMap<>();
        map.put('A', 0);
        map.put('C', 1);
        map.put('G', 2);
        map.put('T', 3);
        List<String> res = new ArrayList<>();
        Map<Integer, Integer> count = new HashMap<>();
        int cur = 0;
        for (int i = 0; i < 10; ++i) {
            cur <<= 2;
            cur += map.get(s.charAt(i));
        }
        count.put(cur, 1);
        for (int i = 10; i < s.length(); ++i) {
            cur = ((cur << 2) & ((1 << 20) - 1)) + map.get(s.charAt(i));
            count.put(cur, count.getOrDefault(cur, 0) + 1);
            if (count.get(cur) == 2) {
                res.add(s.substring(i - 9, i + 1));
            }
        }
        return res;
    }
}
